components/raftstore/src/store/worker/snap_gen.rs

```diff
@@ -0,0 +1,268 @@
+    }
+}
+
+pub struct Runner<EK, R, T>
+where
+    EK: KvEngine,
+    T: PdClient + 'static,
+{
+    tiflash_stores: HashMap<u64, bool>,
+
+    engine: EK,
+    mgr: SnapManager,
+    router: R,
+    pd_client: Option<Arc<T>>,
+    pool: FuturePool,
+}
+
+impl<EK, R, T> Runner<EK, R, T>
+where
+    EK: KvEngine,
+    R: CasualRouter<EK>,
+    T: PdClient + 'static,
+{
+    pub fn new(
+        engine: EK,
+        mgr: SnapManager,
+        router: R,
+        pd_client: Option<Arc<T>>,
+        pool: FuturePool,
+    ) -> Runner<EK, R, T> {
+        Runner {
+            tiflash_stores: HashMap::default(),
+            engine,
+            mgr,
+            router,
+            pd_client,
+            pool,
+        }
+    }
+}
+
+impl<EK, R, T> Runnable for Runner<EK, R, T>
+where
+    EK: KvEngine,
+    R: CasualRouter<EK> + Send + Clone + 'static,
+    T: PdClient,
+{
+    type Task = Task<EK::Snapshot>;
+
+    fn run(&mut self, task: Task<EK::Snapshot>) {
+        match task {
+            Task::Gen {
+                region_id,
+                last_applied_term,
+                last_applied_state,
+                kv_snap,
+                canceled,
+                notifier,
+                for_balance,
+                to_store_id,
+            } => {
+                // It is safe for now to handle generating and applying snapshot concurrently,
+                // but it may not when merge is implemented.
+                let mut allow_multi_files_snapshot = false;
+                // if to_store_id is 0, it means the to_store_id cannot be found
+                if to_store_id != 0 {
+                    if let Some(is_tiflash) = self.tiflash_stores.get(&to_store_id) {
+                        allow_multi_files_snapshot = !is_tiflash;
+                    } else {
+                        let is_tiflash = self.pd_client.as_ref().map_or(false, |pd_client| {
+                            if let Ok(s) = pd_client.get_store(to_store_id) {
+                                return s.get_labels().iter().any(|label| {
+                                    label.get_key().to_lowercase() == ENGINE
+                                        && label.get_value().to_lowercase() == TIFLASH
+                                });
+                            }
+                            true
+                        });
+                        self.tiflash_stores.insert(to_store_id, is_tiflash);
+                        allow_multi_files_snapshot = !is_tiflash;
+                    }
+                }
+                SNAP_COUNTER.generate.all.inc();
+                let ctx = SnapGenContext {
+                    engine: self.engine.clone(),
+                    mgr: self.mgr.clone(),
+                    router: self.router.clone(),
+                    start: UnixSecs::now(),
+                };
+                let scheduled_time = Instant::now_coarse();
+                self.pool.spawn(async move {
```

Comment:
Do we have a limit on the concurrent gen tasks being spawned?

Reply:
Good point. It doesn't look like we have one. Many snap gen tasks may run at the same time and they would be thrown away in src/server/snap.rs if they exceed `concurrent_send_snap_limit`, which will be wasted work:
```
if self.sending_count.load(Ordering::SeqCst) >= self.cfg.concurrent_send_snap_limit
{
    warn!(
        "too many sending snapshot tasks, drop Send Snap[to: {}, snap: {:?}]",
        addr, msg
    );
    cb(Err(Error::Other("Too many sending snapshot tasks".into())));
    return;
}
```
Maybe we should add a limit based on `concurrent_send_snap_limit`? We can set the `max_tasks` of the `FuturePool`.

Reply:
I added a semophore to limit the concurrency of snap gen tasks. I'm using `tokio::sync::Semaphore` which seems to be the only semaphore library used in the TiKV codebase. The current approach doesn't limit the number of spawned threads but it can control the number of threads that are doing actual work. Would this be an acceptable approach?

Reply:
Seems that the semaphore can control the parallelism but cannot control the concurrency? `SNAP_GENERATOR_MAX_POOL_SIZE ` has already been enforced by the thread pool, so this semaphore does not really have any effects?

Reply:
The other thing that I'm a bit hesitate about is the capacity of the semaphore. I'm not sure if running N snap gen tasks is able to fully utilize a pool with N threads, since the snap gen work involves IO heavily. Wdyt? @Connor1996

Reply:
If `ctx.handle_gen()` is async,  semaphore kinda make sense, however, it is a sync call. I think the thread pool size and semaphore here are equivalent, aren't they? If you worried about IO being the bottleneck, then reducing the thread pool size would also help?

Reply:
With a second thought, `ctx.handle_gen()` is a sync call on IO, so `SNAP_GENERATOR_MAX_POOL_SIZE` should already control the parallelism.

Reply:
OK, I didn't realize that a sync call would prevent scheduling in the YATP thread pool. Now it looks like the semaphore is unnecessary and I'll get it removed. @v01dstar Thanks for pointing it out earlier!